core/drand_beacon.go

```diff
@@ -69,108 +71,59 @@ type Drand struct {
 	}
 
-	// trick to always set the listening address by default based on the
-	// identity. If there is an option to set the address, it will override the
-	// default set here..
-	d := &Drand{
-		store:   s,
-		priv:    priv,
-		opts:    c,
-		log:     logger,
-		exitCh:  make(chan bool, 1),
-		version: common.GetAppVersion(),
-	}
-	if err := setupDrand(d, c); err != nil {
-		return nil, err
-	}
-	return d, nil
-}
-
-func setupDrand(d *Drand, c *Config) error {
-	// Set the private API address to the command-line flag, if given.
-	// Otherwise, set it to the address associated with stored private key.
-	privAddr := c.PrivateListenAddress(d.priv.Public.Address())
-	pubAddr := c.PublicListenAddress("")
-	// ctx is used to create the gateway below.
-	// Gateway constructors (specifically, the generated gateway stubs that require it)
-	// do not actually use it, so we are passing a background context to be safe.
-	ctx := context.Background()
-	var err error
-	d.log.Infow("", "network", "init", "insecure", c.insecure)
-	if pubAddr != "" {
-		handler, err := http.New(ctx, &drandProxy{d}, c.Version(), d.log.With("server", "http"))
-		if err != nil {
-			return err
-		}
-		if d.pubGateway, err = net.NewRESTPublicGateway(ctx, pubAddr, c.certPath, c.keyPath, c.certmanager, handler, c.insecure); err != nil {
-			return err
-		}
-	}
-	d.privGateway, err = net.NewGRPCPrivateGateway(ctx, privAddr, c.certPath, c.keyPath, c.certmanager, d, c.insecure, d.opts.grpcOpts...)
-	if err != nil {
-		return err
-	}
-	p := c.ControlPort()
-	d.control = net.NewTCPGrpcControlListener(d, p)
-	go d.control.Start()
-	d.log.Infow("", "private_listen", privAddr, "control_port", c.ControlPort(), "public_listen", pubAddr, "folder", d.opts.ConfigFolderMB())
-	d.privGateway.StartAll()
-	if d.pubGateway != nil {
-		d.pubGateway.StartAll()
+	bp := &BeaconProcess{
+		store:       store,
+		log:         log,
+		priv:        priv,
+		version:     version,
+		opts:        opts,
+		privGateway: privGateway,
+		pubGateway:  pubGateway,
+		control:     control,
+		exitCh:      make(chan bool, 1),
 	}
-	return nil
+	return bp, nil
 }
 
 // LoadDrand restores a drand instance that is ready to serve randomness, with a
 // pre-existing distributed share.
-func LoadDrand(s key.Store, c *Config) (*Drand, error) {
-	d, err := initDrand(s, c)
-	if err != nil {
-		return nil, err
-	}
-	d.group, err = s.LoadGroup()
+func (bp *BeaconProcess) Load() (*BeaconProcess, error) {
+	var err error
+
+	bp.group, err = bp.store.LoadGroup()
 	if err != nil {
 		return nil, err
 	}
-	checkGroup(d.log, d.group)
-	d.share, err = s.LoadShare()
+
+	checkGroup(bp.log, bp.group)
+	bp.share, err = bp.store.LoadShare()
 	if err != nil {
 		return nil, err
 	}
-	d.log.Debugw("", "beacon_id", d.group.ID, "serving", d.priv.Public.Address())
-	d.dkgDone = true
-	return d, nil
+
+	bp.log.Debugw("", "beacon_id", bp.group.ID, "serving", bp.priv.Public.Address())
+	bp.dkgDone = true
+
+	return bp, nil
```

Comment:
No need to return "bp" here since you're mutating the struct anyway - and you're not using the return value anywhere.

Reply:
Fixed on a previous commit.